### PR TITLE
Add online offline status

### LINKED ISSUE #63: Online/offline status in backend
..based on whether any socket of the user has recently pinged.

(ping @heyitsmeuralex.. no pun intended :package:)

### PR DESCRIPTION
**Pull request by towerofnix**
Includes #62! Merge that one before this.

Fixes #63.

This adds a simple online/offline status system. Any serialized user will include whether the user is online or not (in the `online` field). In order to do this, I added a `sessionID` field to socket data; this is filled in by the client's `pong data` event, which is sent in response to the server's `ping for data` event. (`ping for data` is sent as soon as the socket connects, and on the ping interval.) In order to see if a user is online or not, the API simply checks for any socket whose session ID belongs to that user.

### CODE REVIEW

**Comment by towerofnix** [comment]
On second look this doesn't actually *do* anything with the commit in #62. Guess I shoudln't have branched from that, hm?

File: api.js [hunk 1]
```diff
@@ -46,6 +46,17 @@ module.exports = async function attachAPI(app, {wss, db}) {
...
+  const isUserOnline = async function(userID) {
+    // Simple logic: a user is online iff there is at least one socket whose
+    // session belongs to that user.
+
+    const sessions = await db.sessions.find({userID})
+
+    return Array.from(connectedSocketsMap.values())
+      .some(socketData => sessions
+        .some(session => session._id === socketData.sessionID))
+  }
```
Comment (heyitsmeuralex) [inline api.js]: not webscale enough tbh

File: api.js [hunk 2]
```diff
@@ -770,25 +783,46 @@ module.exports = async function attachAPI(app, {wss, db}) {
           return
         }
 
-        connectedSocketsMap.set(socket, {
+        Object.assign(connectedSocketsMap.get(socket), {
           channelID: data // channelID
         })
+      } else if (evt === 'pong data') {
+        // Not the built-in pong; this event is used for gathering
+        // socket-specific data.
```
Comment (heyitsmeuralex) [inline api.js]: Hey, uh, `socket.ping` can take a data argument. Check out the ws documentation.

Comment (towerofnix) [reply api.js]: I didn't see a way to detect the actual ping event from the browser, though? https://stackoverflow.com/a/45929287/4633828

Comment (heyitsmeuralex) [reply api.js]: Ok!

**Pull request merged**